src/printer_tree.c

```diff
@@ -0,0 +1,3107 @@
+};
+
+/******************************************************************************
+ * Read getters
+ *****************************************************************************/
+
+/**
+ * @brief Functions that do not change the state of the tree_structure.
+ *
+ * For details see trt_fp_modify_ctx.
+ */
+struct trt_fp_read {
+    struct trt_keyword_stmt (*module_name)(const struct trt_tree_ctx *);                        /**< Get name of the module. */
+    struct trt_node (*node)(struct trt_parent_cache, const struct trt_tree_ctx *);              /**< Get current node. */
+    ly_bool (*if_sibling_exists)(const struct trt_tree_ctx *);  /**< Check if node's sibling exists. */
+};
+
+/******************************************************************************
+ * All getters
+ *****************************************************************************/
+
+/**
+ * @brief A set of all necessary functions that must be provided for the printer.
+ */
+struct trt_fp_all {
+    struct trt_fp_modify_ctx modify;    /**< Function pointers which modify state of trt_tree_ctx. */
+    struct trt_fp_read read;            /**< Function pointers which only reads state of trt_tree_ctx. */
+    struct trt_fp_print print;          /**< Functions pointers for printing special items in node. */
+};
+
+/******************************************************************************
+ * Printer context
+ *****************************************************************************/
+
+/**
+ * @brief Main structure for trp component (printer part).
+ */
+struct trt_printer_ctx {
+    struct ly_out *out;     /**< Handler to printing. */
+    struct trt_fp_all fp;   /**< 'tro' functions callbacks. */
+    size_t max_line_length; /**< The maximum number of characters that can be
+                               printed on one line, including the last. */
+};
+
+/******************************************************************************
+ * Tro functions
+ *****************************************************************************/
+
+/**
+ * @brief The name of the section to which the node belongs.
+ */
+typedef enum {
+    TRD_SECT_MODULE = 0,    /**< The node belongs to the "module: <module_name>:" label. */
+    TRD_SECT_AUGMENT,       /**< The node belongs to some "augment <target-node>:" label. */
+    TRD_SECT_RPCS,          /**< The node belongs to the "rpcs:" label. */
+    TRD_SECT_NOTIF,         /**< The node belongs to the "notifications:" label. */
+    TRD_SECT_GROUPING,      /**< The node belongs to some "grouping <grouping-name>:" label. */
+    TRD_SECT_YANG_DATA      /**< The node belongs to some "yang-data <yang-data-name>:" label. */
+} trt_actual_section;
+
+/**
+ * @brief Ancestors who have a strong influence on their children
+ */
+typedef enum {
+    TRD_ANCESTOR_ELSE = 0,
+    TRD_ANCESTOR_RPC_INPUT,
+    TRD_ANCESTOR_RPC_OUTPUT,
+    TRD_ANCESTOR_NOTIF
+} trt_ancestor_type;
```

Comment:
I just want to add that you can detect all of these ancestors directly from `struct lysc_node` (flags `LYS_IS_INPUT`, `LYS_IS_OUTPUT`, `LYS_IS_NOTIF`). But not from `struct lysp_node` so I am not sure whether that helps in any way.